net: ip: ipv4: Add fragmented packet support

thedjnK opened the pull request:
Adds support for incoming and outgoing IPv4 fragmented packet support, allowing a single packet that is too large to be sent to be split up and sent successfully.

Fixes #49783

Tested using `smp_svr` sample on `native_posix_64` and a Cortex M board with ethernet, test passes with `native_posix_64` on local PC.

thedjnK committed:
net: ipv4: Add fragmented packet support

Adds support for incoming and outgoing IPv4 fragmented packet support,
allowing a single packet that is too large to be sent to be split up
and sent successfully.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

rlubos commented on subsys/net/ip/ipv4_fragment.c:
```diff
@@ -0,0 +1,647 @@
+#define NET_BUF_TIMEOUT K_MSEC(100)
+
+/* How long to wait for a buffer */
+#define FRAG_BUF_WAIT K_MSEC(10)
```
It seems that this define is not used anywhere?

rlubos reviewed (commented):
Thank you for the effort, the implementation seems pretty neat. Do you plan to add some basic tests for the IPv4 fragmentation? Currently, the code wouldn't even be built by the CI, not even mentioning possible future regressions.

nordicjm commented on include/zephyr/net/net_pkt.h:
```diff
@@ -234,6 +234,11 @@ struct net_pkt {
 #endif /* CONFIG_NET_VLAN */
 
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
+	uint16_t ipv4_fragment_flags;	/* Fragment offset and M (More Fragment) flag */
+	uint16_t ipv4_fragment_id;	/* Fragment ID */
```
Will have a play.

Not sure on what to do for testing at present, the one idea I do have is testing it using mcumgr to send a very long echo and checking it is received and sent back, will need to look at the network tests to see if this is possible and how they work though.

rlubos replied:
There is an existing test suite for IPv6 fragmentation:
https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/net/ipv6_fragment

It's not a rule of thumb, but typically network tests either create a dummy interface to capture raw IP packets or loopback for local communication.

thedjnK replied:
Fixed

tbursztyka reviewed (commented):
would be nice to get a commit adding test for this feature (as it has been done for ipv6 fragment support).

thedjnK commented on subsys/net/ip/ipv4_fragment.c:
```diff
@@ -0,0 +1,647 @@
+		}
+
+		offset = net_pkt_ipv4_fragment_offset(pkt);
+
```
The line gap is how it is in other files, e.g. ipv6_fragment.c

thedjnK commented on include/zephyr/net/net_pkt.h:
```diff
@@ -234,6 +234,11 @@ struct net_pkt {
 	uint16_t vlan_tci;
 #endif /* CONFIG_NET_VLAN */
 
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
```
Fixed

thedjnK commented on subsys/net/ip/net_core.c:
```diff
@@ -71,11 +72,12 @@ static inline enum net_verdict process_data(struct net_pkt *pkt,
+	 * then do not pass it to L2 as the packet does not have link layer headers in it.
 	 */
-	if (IS_ENABLED(CONFIG_NET_IPV6_FRAGMENT) && net_pkt_ipv6_fragment_start(pkt)) {
+	if (IS_ENABLED(CONFIG_NET_IPV4_FRAGMENT) && net_pkt_ipv4_fragment_more(pkt)) {
```
Fixed

thedjnK commented on subsys/net/ip/net_core.c:
```diff
@@ -448,6 +450,10 @@ static inline void l3_init(void)
 	net_icmpv6_init();
 	net_ipv6_init();
 
+	if (IS_ENABLED(CONFIG_NET_IPV4_FRAGMENT)) {
```
Fixed

thedjnK commented:
Will have to make some time to create tests for this

jukkar commented on include/zephyr/net/net_pkt.h:
```diff
@@ -234,6 +234,22 @@ struct net_pkt {
 #endif /* CONFIG_NET_VLAN */
 
+	union {
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
```
Do we need the first part of the check here, wouldn't `#if defined(CONFIG_NET_IPV4_FRAGMENT)` be enough as the fragment symbol is not defined if IPv4 is not enabled.
Same comment for IPv6 check.

thedjnK replied:
Fixed

jukkar commented:
Minor nit about the commit subject, it would be enough to set it to
`net: ipv4: Add fragmented packet support `

thedjnK committed:
tests: net: Add ipv4_fragment test

This adds a test for the new ipv4_fragment feature. This test suite
checks TCP and UDP messages are correctly fragmented and reassembled,
checks that an ICMP error is sent if not all fragments were received
and checks that packets with the do not fragment bit are not
fragmented.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

nordicjm commented:
Tests are about 33% complete thus far, will aim to get them finished and submitted to PR within the next week.

thedjnK committed:
doc: release: 3.3: Add note on IPv4 fragmentation support

Adds a note that IPv4 fragmentation support is now available.

Signed-off-by: Jamie McCrae <spam@helper3000.net>

thedjnK commented on subsys/net/ip/net_if.c:
```diff
@@ -484,6 +485,12 @@ enum net_verdict net_if_send_data(struct net_if *iface, struct net_pkt *pkt)
 		verdict = net_ipv6_prepare_for_send(pkt);
 	}
 
+#if defined(CONFIG_NET_IPV4) && defined(CONFIG_NET_IPV4_FRAGMENT)
```
Fixed

thedjnK commented:
@jukkar @tbursztyka @rlubos Updated with corrections and test/release notes added. There is one outstanding issue with the test which is to check the ID, and since this is generated randomly I'm not sure I can get it from the test to verify it, other than that it's all ready for a review.

jukkar commented on tests/net/ipv4_fragment/src/main.c:
```diff
@@ -0,0 +1,886 @@
+		net_pkt_cursor_init(recv_pkt);
+		net_pkt_skip(recv_pkt, 6);
+		net_pkt_read_be16(recv_pkt, &offset);
+		offset &= 0b1111111111111;
```
Hmm, this does not look like a standard C but is only available as a gcc extension. I suggest that normal hex value is used here instead.

nordicjm replied:
Fixed

jukkar reviewed (commented):
LGTM, only comment about the bit constant that imho should be changed to hex constant.

rlubos reviewed (approved):
LGTM
